toml/src/main/kotlin/org/rust/toml/completion/LocalCargoTomlDependencyCompletionProvider.kt

```diff
@@ -58,32 +61,101 @@ class LocalCargoTomlDependencyCompletionProvider : TomlKeyValueCompletionProvide
     }
 
     override fun completeValue(keyValue: TomlKeyValue, result: CompletionResultSet) {
-        val name = CompletionUtil.getOriginalElement(keyValue.key)?.text ?: return
+        val keySegment = keyValue.key.segments.singleOrNull() ?: return
+        val name = CompletionUtil.getOriginalElement(keySegment)?.text ?: return
 
         val crate = try {
             CratesLocalIndexService.getInstance().getCrate(name)
         } catch (e: CratesLocalIndexException) {
             return
         }
         val sortedVersions = crate?.sortedVersions ?: return
+        val elements = makeVersionCompletions(sortedVersions, keyValue)
+        result.withRelevanceSorter(versionsSorter).addAllElements(elements)
+    }
+}
+
+class LocalCargoTomlSpecificDependencyHeaderCompletionProvider : CompletionProvider<CompletionParameters>() {
+    override fun addCompletions(parameters: CompletionParameters, context: ProcessingContext, result: CompletionResultSet) {
+        val key = parameters.position.parent as? TomlKeySegment ?: return
+        val crateNames = try {
+            CratesLocalIndexService.getInstance().getAllCrateNames()
+        } catch (e: CratesLocalIndexException) {
+            return
+        }
 
-        val elements = sortedVersions.mapIndexed { index, variant ->
-            val lookupElement = LookupElementBuilder.create(variant.version)
-                .withInsertHandler(StringValueInsertionHandler(keyValue))
-                .withTailText(if (variant.isYanked) " yanked" else null)
+        val elements = crateNames.map { variant ->
+            LookupElementBuilder.create(variant)
+                .withIcon(AllIcons.Nodes.PpLib)
+                .withInsertHandler { ctx, item ->
+                    val table = key.ancestorStrict<TomlTable>() ?: return@withInsertHandler
+                    if (table.entries.isEmpty()) {
+                        ctx.document.insertString(
+                            ctx.selectionEndOffset + 1,
+                            "\nversion = \"\""
+                        )
 
-            PrioritizedLookupElement.withPriority(
-                lookupElement,
-                index.toDouble()
-            )
+                        EditorModificationUtil.moveCaretRelatively(ctx.editor, 13)
+                        AutoPopupController.getInstance(ctx.project).scheduleAutoPopup(ctx.editor)
+                    }
+                }
+        }
+
+        result.addAllElements(elements)
+    }
+}
+
+class LocalCargoTomlSpecificDependencyVersionCompletionProvider : TomlKeyValueCompletionProviderBase() {
+    override fun completeKey(keyValue: TomlKeyValue, result: CompletionResultSet) {
+        result.addElement(
+            LookupElementBuilder.create("version")
+                .withInsertHandler { ctx, item ->
```

Comment:
`Duplicated code fragment (14 lines long)`
Please try to reduce code duplication in all these providers